src/cargo/util/toml/mod.rs

```diff
@@ -1578,10 +1578,30 @@ impl TomlManifest {
         let cargo_features = me.cargo_features.as_ref().unwrap_or(&empty);
         let features = Features::new(cargo_features, config, &mut warnings, source_id.is_path())?;
 
-        let project = me.project.clone().or_else(|| me.package.clone());
-        let project = &mut project.ok_or_else(|| anyhow!("no `package` section found"))?;
+        let mut package = match (&me.package, &me.project) {
+            (Some(_), Some(project)) => {
+                config.shell().warn(format!(
+                    "manifest at `{}` contains both `project` and `package`, \
+                    this could become a hard error in the future",
+                    package_root.display()
+                ))?;
```

Comment:
Should we also avoid emitting error against non-local packages here?

In addition, we could have tests to ensure that we only emit errors on local packages.